### PR TITLE
Add mm.MIDIPlayer

### LINKED ISSUE #107: Add WebMIDI support for Player

### PR DESCRIPTION
**Pull request by notwaldorf**
Fixes #107

### CODE REVIEW

File: music/src/core/player.ts [hunk 1]
```diff
@@ -541,3 +541,134 @@ export class PlayerWithClick extends Player {
...
+ * If you want to use a particular MIDI output, you must update the `output`
+ * property before calling `start`, otherwise a message will be sent to
+ * all connected MIDI outputs:
+ *
+ * Example (easy mode):
+ *
+ *   `
+ *    const player = new mm.MIDIPlayer();
+ *    player.requestMIDIAccess().then(() => {
+ *      player.output = [player.midiOutputs[0]];
```
Comment (adarob) [inline music/src/core/player.ts]: does this only play on one of the outputs?

File: music/src/core/player.ts [hunk 2]
```diff
@@ -541,3 +541,134 @@ export class PlayerWithClick extends Player {
...
+ *
+ * You can also explicitly request MIDI access outside of the player, in
+ * your application, and just update the `output` property before playing:
+ *
+ * Example (advanced mode):
+ *
+ *   `
+ *    navigator.requestMIDIAccess().then((midi) => {
+ *       // Get whichever MIDI outputs you want.
+ *       const outputs = initOutputs(midi);
```
Comment (adarob) [inline music/src/core/player.ts]: what is initOutputs?

File: music/src/core/player.ts [hunk 3]
```diff
@@ -541,3 +541,134 @@ export class PlayerWithClick extends Player {
...
+ * property before calling `start`, otherwise a message will be sent to
+ * all connected MIDI outputs:
+ *
+ * Example (easy mode):
+ *
+ *   `
+ *    const player = new mm.MIDIPlayer();
+ *    player.requestMIDIAccess().then(() => {
+ *      player.output = [player.midiOutputs[0]];
+ *      player.start();
```
Comment (adarob) [inline music/src/core/player.ts]: don't you have to send a NoteSequence to start?

File: music/src/core/player.ts (hunk 1 above)
Comment (notwaldorf) [reply music/src/core/player.ts]: in this example, yes. it's an array though, so you can pass a number of outputs (like in the example below)

File: music/src/core/player.ts (hunk 2 above)
Comment (notwaldorf) [reply music/src/core/player.ts]: some code that you would write to initialize all the midi outputs. should i write it all out you think?

File: music/src/core/player.ts (hunk 1 above)
Comment (adarob) [reply music/src/core/player.ts]: can you add to the end of the last line `# arbitrarily choose first port`

Comment (adarob) [reply music/src/core/player.ts]: actually, isn't it easier (for easy mode) just to send it to all ports?

File: music/src/core/player.ts [hunk 4]
```diff
@@ -541,3 +541,134 @@ export class PlayerWithClick extends Player {
...
 }
+
+/**
+ * A `NoteSequence` player that uses a MIDI output for playing. Note that
+ * WebMIDI is not supported in all browsers. A
+ * [polyfill](https://cwilso.github.io/WebMIDIAPIShim/) exists, but it too
+ * requires a plugin to be installed on the user's computer, so it might not
+ * work in all cases.
+ *
+ * If you want to use a particular MIDI output, you must update the `output`
```
Comment (adarob) [inline music/src/core/player.ts]: s/output/output port/g

File: music/src/core/player.ts [hunk 5]
```diff
@@ -541,3 +541,134 @@ export class PlayerWithClick extends Player {
...
+ *       // Get whichever MIDI outputs you want.
+ *       const outputs = initOutputs(midi);
+ *       player = new mm.MIDIPlayer();
+ *       player.output = [outputs[0], output[1]];
+ *       player.start(seq)
+ *     });`
+ */
+export class MIDIPlayer extends BasePlayer {
+  public output: WebMidi.MIDIOutput[] = [];
+  public readonly midiOutputs: WebMidi.MIDIOutput[] = [];
```
Comment (adarob) [inline music/src/core/player.ts]: s/midiOutputs/availableOutputs/g?

File: music/src/core/player.ts [hunk 6]
```diff
@@ -541,3 +541,134 @@ export class PlayerWithClick extends Player {
...
+ *    navigator.requestMIDIAccess().then((midi) => {
+ *       // Get whichever MIDI outputs you want.
+ *       const outputs = initOutputs(midi);
+ *       player = new mm.MIDIPlayer();
+ *       player.output = [outputs[0], output[1]];
+ *       player.start(seq)
+ *     });`
+ */
+export class MIDIPlayer extends BasePlayer {
+  public output: WebMidi.MIDIOutput[] = [];
```
Comment (adarob) [inline music/src/core/player.ts]: s/output/outputs/g?

File: music/src/core/player.ts (hunk 2 above)
Comment (adarob) [reply music/src/core/player.ts]: What does it mean to "initialize" the outputs?

Comment (notwaldorf) [reply music/src/core/player.ts]: like...the bit that does `navigator.requestMidiAccess()` and then iterates through all the midiAccess.outputs to get the values

Comment (adarob) [reply music/src/core/player.ts]: isn't navigator.requestMidiAccess already being called?

Comment (notwaldorf) [reply music/src/core/player.ts]: the bit where you loop through outputs. i'll make this explicit.

File: music/src/core/player.ts (hunk 1 above)
Comment (notwaldorf) [reply music/src/core/player.ts]: sure!

File: music/src/core/player.ts [hunk 7]
```diff
@@ -541,3 +541,146 @@ export class PlayerWithClick extends Player {
...
+ * requires a plugin to be installed on the user's computer, so it might not
+ * work in all cases.
+ *
+ * If you want to use a particular MIDI output port, you must update the
+ * `output` property before calling `start`, otherwise a message will be sent to
+ * all connected MIDI outputs:
+ *
+ * Example (easy mode):
+ *
+ *   `
```
Comment (adarob) [inline music/src/core/player.ts]: is this the right way to do it or should it be ``` for markdown?

Comment (notwaldorf) [reply music/src/core/player.ts]: great question! i copied it from the SoundfontPlayer docs, but lemme build the docs and see if it's borked

**Pull request merged**